src/app/login/login.component.ts

```diff
@@ -10,34 +11,45 @@ import {ActivatedRoute, Router} from '@angular/router';
 })
 export class LoginComponent implements OnInit {
 
-  constructor(private authService: AuthenticationService, private router: Router, private route: ActivatedRoute) { }
-
-  username: string;
-  password: string;
+  loginForm = this.fb.group({
+    email: ['', Validators.required],
+    password: ['', Validators.required],
+  });
   returnUrl: string;
-  showSpinner: boolean;
-  showLoginError: boolean;
+  loading = false;
+  showLoginErrorMessage = false;
+
+  constructor(
+    private authService: AuthenticationService,
+    private router: Router,
+    private route: ActivatedRoute,
+    private fb: FormBuilder,
+  ) { }
 
   ngOnInit() {
     // get return url from route parameters or default to '/'
     this.returnUrl = this.route.snapshot.queryParams['returnUrl'] || '/';
-    this.showSpinner = false;
   }
 
-  login(): void {
-    this.showLoginError = false;
-    this.showSpinner = true;
-
-    this.authService.login(this.username, this.password)
+  onLoginSubmit(): void {
+    if (!this.loginForm.valid) {
+      return;
+    }
+    // indicate loading so the form can be disabled and loading animations shown
+    this.loading = true;
+    // using AuthService send a login request with user email and password
+    this.authService.login(this.loginForm.get('email').value, this.loginForm.get('password').value)
```

Comment:
this is ok, but for larger forms it's better to simply use loginForm.value and pass the object to the service like that

Reply:
it would be {email: "smth", password: "admin"}